### ISSUE TITLE
GTK2 obsolete and being discontinued

### ISSUE THREAD

**Issue by manuelafm** [K-3D/k3d#30]
Hello Bart,

This package depends on GTK2 libraries, in particular gtksourceview2.

In Debian gtksourceview2 is going to be removed in the next few months/year (see below), so we have to doe something about it, either port it or disable the dependency and miss functionality if it's an optional dependency, or in the worst case remove the pacakge.

Are there any plans to update GTK dependencies to v3 or even v4?  Or moving to Qt?  I guess not, specially the latter, because this project seems to be in maintenance mode for many years.

For reference, this is the report: https://bugs.debian.org/885686

> As announced [1], we do not intend to release Debian 10 "Buster" with
> the old libgnome (and related) libraries. These libraries have been
> deprecated and unmaintained for several years.
>
> Your package depends on gtksourceview2 which has not had a new release
> since 2010.
> 
> Please port your package to GTK3 and gtksourceview3 and related
> maintained libraries. Otherwise, please consider requesting that your
> package be removed from Debian to help us complete this goal.

**Comment by jbicha** [comment]
I don't know of any apps using gtk4 yet, so it's too early to worry about gtk4.

**Comment by barche** [comment]
Hi,

I think the sourceview plays a minor role, so that should be fixable. However, does this also affect libgtk and libgtkmm?

**Comment by manuelafm** [comment]
I don't know exactly, sooner or later it will do, since they've been unmaintained upstream for 7 years.

I don't know if the GNOME team in Debian know it either at this point an exact date, it's not easy to notify hundreds of packages depending on these libraries, but I think that the plan is to remove them within this release cycle before the preparations for the next stable, so in 9~12 months if not earlier.

Maybe @jbicha will chime in to share more data :)

**Comment by jbicha** [comment]
gtk2 and gtkmm2 will still be in Debian 10 "Buster" which is scheduled for release in 2019. About 2 years after that, Debian 11 "Bullseye" is expected. Debian 11 might drop gtk2 (or it might not; it's a bit too early to say).

gtk2 has only been unmaintained in my opinion for a year (its last release was September 2016 when gtk 3.22 was announced as the new Long Term Support branch). But several libraries commonly used in gtk2 apps have been unmaintained since GNOME 3 was released in 2011. It's those libraries that we're trying to remove from Buster pretty soon.

**Comment by barche** [comment]
It seems like the logical way forward is to move to gtksourceview3, which would also mean porting everything to gtk(mm)3.... Can't promise anything on the timing for this, but the GTK migration guide looks very helpful.

**Comment by Megalomaniak** [comment]
Considering the name I actually expected this program to be part of the KDE suite and thus based on QT to be honest. Since porting might be necessary anyways might this be considered?

for reference going with QT and pyQT would make K-3D more in line with VFX Platform standardization:
https://www.vfxplatform.com/

> The VFX Reference Platform is a set of tool and library versions to be used as a common target platform for building software for the VFX industry. Its purpose is to minimise incompatibilities between different software packages, ease the support burden for Linux-based pipelines and encourage further adoption of Linux by software vendors. The Reference Platform is updated annually by a group of software vendors in collaboration with the [Visual Effects Society Technology Committee](https://www.visualeffectssociety.com/ves-committees?jump-9).

**Comment by barche** [comment]
I ery much prefer Qt (QML more specifically) myself, but for this particular issue I started to bite the bullet and convert to GTK3.

**Comment by Megalomaniak** [comment]
If I may ask how portable is GTK3? I recall GTK2 having been a real pain for deploying and maintaining windows ports of GIMP for an example, which is not to say that it didn't get done by third-parties anyways.

**Comment by barche** [comment]
Yes, it is the main reason the Windows build is so far behind. I still need to test it, but it appears things are much better now with [MSYS2](https://www.msys2.org/), which also has packages for GTK(mm)3. So gently porting K-3D to GTK3 seems like the road of least resistance to ensure its survival with a very limited time budget.

Regarding Qt, I did start a little bit on a QML port in my fork: https://github.com/barche/k3d
Unfortunately, due to other priorities I have no idea when or if ever I will pick up that work again.

**Comment by Megalomaniak** [comment]
Right, if GTK3 will work out fine enough there is no point dedicating effort towards porting to Qt.

**Comment by manuelafm** [comment]
@barche , any progress on this update?  It seems that in the short term we have to disable this functionality in Debian because gtksourceview2 is going to be removed imminently.

**Comment by barche** [comment]
It is turning out to be more work than I anticipated, so I think just disabling the source view is the correct action if it's urgent.

**Comment by manuelafm** [comment]
@barche I suspect that more GTK-2 era stuff will be removed soon from Debian, is there any progress in the migration?

**Comment by jiapei100** [comment]
Expecting **gtk3** anyway....

**Comment by barche** [comment]
No progress since a long time. I did push what I have so far to the gtk3 branch, but it is incomplete (doesn't compile yet). It is a lot more work than I had anticipated, so anyone is welcome to pick it up.
